DOC: Add example to rv_continuous.fit

m-maggi opened the pull request:
#### Reference issue
Closes gh-19225 

#### What does this implement/fix?
Extend the documentation of scipy.stats.rv_continuous.fit

m-maggi committed:
DOC: Add example to rv_continuous.fit

See Issue #19225

lucascolley commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -2589,6 +2589,16 @@ def fit(self, data, *args, **kwds):
+
+        Using a custom optimizer:
+
+        >>> def optimizer_NM(func, x0=[0.6, 1.6], args=(), disp=0):
```
where did you take these argument values from / why did you choose these values?

m-maggi replied:
Thanks for reviewing. I replaced the normal example with the beta, as normal does not use the ``optimizer`` argument https://github.com/m-maggi/scipy/blob/f964b3c812eb911a18f30eb26c146108e020862e/scipy/stats/_continuous_distns.py#L393-L399

Furthermore I added 
> The optimizer must return the estimated parameters.

In the doctoring.

lucascolley reviewed (commented):
Looks good so far - could you also address this part of https://github.com/scipy/scipy/issues/19225#issuecomment-1714637380?

> It doesn't state what the optimizer should return or guarantee that user guesses will be respected.

m-maggi committed:
DOC: Modify example in rv_continuous.fit.

- Remove previous example with stats.norm
- Add example using stats.beta
- Clarify that ``optimizer`` must return the estimated parameters.

m-maggi committed:
DOC: Change method in ``minimize``

j-bowhay commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -2559,12 +2560,26 @@ def fit(self, data, *args, **kwds):
+
+        >>> from scipy.optimize import minimize
+        >>> def custom_optimizer(func, x0, args=(), disp=0):
+        ...     res = minimize(func, x0, args, method="COBYLA")
```
```suggestion
        ...     res = minimize(func, x0, args, method="COBYLA", disp=disp)
```

m-maggi replied:
thanks for reviewing, I will actually remove the disp argument from the docstring as it is anyways set to 0 when calling fit in stats/_distn_infrastructure.py
https://github.com/scipy/scipy/blob/1004bb958b43426cd6cfd5cdf5baafa4f138902c/scipy/stats/_distn_infrastructure.py#L2640C57-L2640C57

```python3
vals = optimizer(func, x0, args=(data,), disp=0)
obj = func(vals, data)
```

m-maggi committed:
DOC: Modify rv_continuous.fit docstring

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -2500,8 +2500,9 @@ def fit(self, data, *args, **kwds):
-              function to be optimized) and ``disp=0`` to suppress
-              output as keyword arguments.
+              function to be optimized) and ``disp``. 
+              The ``fit`` method calls always the optimizer with ``disp=0``.
```
```suggestion
              The ``fit`` method always calls the optimizer with ``disp=0``.
```
or just 
```suggestion
              The ``fit`` method calls the optimizer with ``disp=0`` to suppress output.
```

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -2559,12 +2560,26 @@ def fit(self, data, *args, **kwds):
         >>> from scipy.stats import beta
         >>> a, b = 1., 2.
-        >>> x = beta.rvs(a, b, size=1000)
+        >>> x = beta.rvs(a, b, size=1000, random_state=80)
```
Elsewhere in the docs there has been a push to use `default_rng` with a larger seed to encourage use of `Generator`s and avoid repeated use of the same seed.

mdhaber commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -2559,12 +2560,26 @@ def fit(self, data, *args, **kwds):
+
+        >>> from scipy.optimize import minimize
+        >>> def custom_optimizer(func, x0, args=(), disp=0):
+        ...     res = minimize(func, x0, args, method="COBYLA", options={"disp": disp})
```
```suggestion
        ...     res = minimize(func, x0, args, method="slsqp", options={"disp": disp})
```
Let's avoid recommending use of COBYLA. SLSQP seems to be much more reliable.

mdhaber reviewed (commented):
Thanks @m-maggi. A few comments.

m-maggi committed:
Merge branch 'scipy:main' into doc/rv_continuous_fit

m-maggi committed:
Update scipy/stats/_distn_infrastructure.py

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

m-maggi committed:
DOC Use default_rng and slsqp
